be/src/exec/vectorized/json_scanner.cpp

```diff
@@ -674,129 +716,389 @@ Status JsonReader::_read_and_parse_json() {
+            return Status::DataQualityError(err_msg.c_str());
+        }
+        return Status::OK();
+    }
+
+    case TYPE_TINYINT: {
+        StringParser::ParseResult parse_result = StringParser::PARSE_SUCCESS;
+        auto i8 = StringParser::string_to_int<int8_t>(sv.data(), sv.length(), &parse_result);
+        if (parse_result == StringParser::PARSE_SUCCESS) {
+            _construct_numeric_column(column, i8);
+        } else {
+            // Attemp to parse the string as float.
+            auto d = StringParser::string_to_float<double>(sv.data(), sv.length(), &parse_result);
+            if (parse_result == StringParser::PARSE_SUCCESS) {
+                i8 = static_cast<int16_t>(d);
+                _construct_numeric_column(column, i8);
+            }
+
+            std::string err_msg = strings::Substitute("Unable to cast string value to TINYINT. value=$0, column=$1",
+                                                      std::string(sv.data(), sv.size()), desc->col_name());
+            return Status::DataQualityError(err_msg.c_str());
+        }
+        return Status::OK();
+    }
+
+    case TYPE_FLOAT: {
+        StringParser::ParseResult parse_result = StringParser::PARSE_SUCCESS;
+        auto f = StringParser::string_to_float<float>(sv.data(), sv.length(), &parse_result);
+        if (parse_result == StringParser::PARSE_SUCCESS) {
+            _construct_numeric_column(column, f);
+        } else {
+            std::string err_msg = strings::Substitute("Unable to cast string value to FLOAT. value=$0, column=$1",
+                                                      std::string(sv.data(), sv.size()), desc->col_name());
+            return Status::DataQualityError(err_msg.c_str());
+        }
+        return Status::OK();
+    }
+
+    case TYPE_DOUBLE: {
+        StringParser::ParseResult parse_result = StringParser::PARSE_SUCCESS;
+        auto d = StringParser::string_to_float<double>(sv.data(), sv.length(), &parse_result);
+        if (parse_result == StringParser::PARSE_SUCCESS) {
+            _construct_numeric_column(column, d);
+        } else {
+            std::string err_msg = strings::Substitute("Unable to cast string value to DOUBLE. value=$0, column=$1",
+                                                      std::string(sv.data(), sv.size()), desc->col_name());
+            return Status::DataQualityError(err_msg.c_str());
+        }
+        return Status::OK();
     }
 
+    // Treat all other types as VARCHAR.
     default: {
-        column->append_nulls(1);
-        break;
+        _construct_string_column(column, Slice{sv.data(), sv.size()});
+        return Status::OK();
     }
     }
 }
 
+void JsonReader::_construct_string_column(Column* col, const Slice& s) {
+    auto column = down_cast<NullableColumn*>(col);
+    auto& data_column = column->data_column();
+    auto& null_column = column->null_column();
+
+    down_cast<BinaryColumn*>(data_column.get())->append(s);
```

Comment:
Based on the calling context of this method, there seems to be no way to guarantee that the column type here is necessarily a BinaryColumn?

Reply:
Corrected. The column type is checked before down_cast now.